google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -372,13 +403,49 @@ void processReceivedMessages(List<ReceivedMessage> messages) {
         // totally expire so that pubsub service sends us the message again.
         continue;
       }
-      outstandingBatch.add(new OutstandingMessage(message, ackHandler));
+      OutstandingMessage outstandingMessage = new OutstandingMessage(message, ackHandler);
+      if (this.exactlyOnceDeliveryEnabled.get()) {
+        outstandingReceipts.put(
+            message.getAckId(), new ReceiptCompleteData(outstandingMessage, false));
+      } else {
+        outstandingBatch.add(outstandingMessage);
+      }
       pendingReceipts.add(ackRequestData);
     }
-
     processBatch(outstandingBatch);
   }
 
+  synchronized void notifyAckSuccess(AckRequestData ackRequestData) {
+
+    if (outstandingReceipts.containsKey(ackRequestData.getAckId())) {
+      ReceiptCompleteData receiptCompleteData = outstandingReceipts.get(ackRequestData.getAckId());
+      OutstandingMessage outstandingMessage = receiptCompleteData.getOutstandingMessage();
+      // Setting to true means that the receipt is complete
+      receiptCompleteData.setReceiptComplete(true);
+      List<OutstandingMessage> outstandingBatch = new ArrayList<>();
+
+      for (Iterator<Entry<String, ReceiptCompleteData>> it =
+              outstandingReceipts.entrySet().iterator();
+          it.hasNext(); ) {
+        Map.Entry<String, ReceiptCompleteData> receipt = it.next();
+        // If receipt is complete then add to outstandingBatch to process the batch
+        if (receipt.getValue().getReceiptComplete()) {
+          it.remove();
+          pendingMessages.putIfAbsent(
```

Comment:
If the message is already contained in pendingMessages, we know it's a duplicate so don't add the message to the outstandingBatch.